src/components/Tooltip/TooltipRenderedOnPageBody/index.native.tsx

```diff
@@ -0,0 +1,137 @@
+import React, {useEffect, useMemo, useRef, useState} from 'react';
```

Comment:
File-scoped comment.

The amount of duplication between `index.tsx` and `index.native.tsx` is unacceptable. This is a big one.

Likely, there's a tough refactoring ahead of us.

Reply:
I'm not saying that we don't need any platform-splitting to `*.tsx` and `*.native.tsx` files, or that we need to implement all complex features on Native that the general-case Web hover-on tooltips need. On Native we'll be showing only the "educational" tooltips.

For example, you correctly assumed that it's not required to dynamically resize tooltips, as educational tootlips likely won't need this behavior soon, or ever. In general, the hover-on tooltips can resize in corner cases.

Still, there are many code de-duplication techniques that can be applied here, even with the simplifications and shortcomings we accepted.

Reply:
Please give this a try. I feel like little effort has been put into de-duplication here so far. But if you run out of ideas, let me know, as I'll likely be able to help.

Reply:
Just a sketch:

```tsx
/**
 * The shared set of properties for the tooltip components
 */
type SharedTooltipProps = {
    /** The text to display in the tooltip. If text is ommitted, only children will be rendered. */
    text?: string;

    /** Maximum number of lines to show in tooltip */
    numberOfLines?: number;

    /** Any additional amount to manually adjust the horizontal position of the tooltip.
     A positive value shifts the tooltip to the right, and a negative value shifts it to the left. */
    shiftHorizontal?: number | (() => number);

    /** Any additional amount to manually adjust the vertical position of the tooltip.
     A positive value shifts the tooltip down, and a negative value shifts it up. */
    shiftVertical?: number | (() => number);

    // ...and other stuff
};

type CommonTooltipContext = {
    isVisible: boolean,
    showTooltip: () => void,
    hideTooltip: () => void,
    updateBounds: (rect: LayoutRectangle) => void;
};

type CommonTooltipProps = SharedTooltipProps & {
    children: (context: CommonTooltipContext) => ReactNode,
};

type TooltipProps = SharedTooltipProps & {
    children: ReactNode;
};

type HoverTooltipProps = TooltipProps & {
    shouldHandleScroll: boolean,
}

/**
 * The common tooltip implementation, leaving the bounds management to its parent
 */
function CommonTooltip(props: CommonTooltipProps): React.JSX.Element {
    throw new Error("TODO: Implement");
    // Most of the existing code goes here
}

/**
 * Web-only general case tooltip shown on hover, able to dynamically react
 * to size and position changes of its child.
 */
function HoverTooltip(
    props: HoverTooltipProps,
    ref: ForwardedRef<BoundsObserver>,
) {
    const {children} = props;

    const target = useRef<HTMLElement | null>(null);
    const initialMousePosition = useRef({x: 0, y: 0});

    const updateTargetAndMousePosition = useCallback((e: MouseEvent) => {
        if (!(e.currentTarget instanceof HTMLElement)) {
            return;
        }
        target.current = e.currentTarget;
        initialMousePosition.current = {x: e.clientX, y: e.clientY};
    }, []);

    const updateTargetPositionOnMouseEnter = useCallback(
        (e: MouseEvent) => {
            updateTargetAndMousePosition(e);
            if (React.isValidElement(children)) {
                children.props.onMouseEnter?.(e);
            }
        },
        [children, updateTargetAndMousePosition],
    );

    return (
        <CommonTooltip {...props}>
            {
                ({isVisible, showTooltip, hideTooltip, updateBounds}) => {
                    const handleBoundsChange = () => {
                        if (!target.current) {
                            return;
                        }
                        // Choose a bounding box for the tooltip to target.
                        // In the case when the target is a link that has wrapped onto
                        // multiple lines, we want to show the tooltip over the part
                        // of the link that the user is hovering over.
                        const betterBounds = chooseBoundingBox(target.current, initialMousePosition.current.x, initialMousePosition.current.y);

                        if (!betterBounds) {
                            return;
                        }

                        updateBounds(betterBounds);
                    };

                    return (
                        <BoundsObserver
                            enabled={isVisible}
                            onBoundsChange={handleBoundsChange}
                            ref={ref}
                        >
                            <Hoverable
                                onHoverIn={showTooltip}
                                onHoverOut={hideTooltip}
                                shouldHandleScroll={props.shouldHandleScroll}
                            >
                                {React.cloneElement(props.children as React.ReactElement, {
                                    onMouseEnter: updateTargetPositionOnMouseEnter,
                                    onLayout: (e: LayoutEvent) => {
                                        // TODO: Figure this out
                                        // if (!shouldRenderWithoutHover) {
                                        //     return;
                                        // }
                                        const childrenRect = e.nativeEvent.target?.getBoundingClientRect();
                                        updateTargetAndMousePosition({
                                            currentTarget: e.nativeEvent.target,
                                            clientX: childrenRect?.left ?? 0,
                                            clientY: childrenRect?.top ?? 0,
                                        } as MouseEvent);
                                        showTooltip();
                                    },
                                })}
                            </Hoverable>
                        </BoundsObserver>
                    );
                }
            }
        </CommonTooltip>
    );
}

/**
 * A tooltip that is displayed over an element to explain it and draw attention to it. Works cross-platform, but might
 * not support all bounds-reacting features of HoverTooltip.
 */
function EducationalTooltip(props: TooltipProps) {
    const {children} = props;

    return (
        <CommonTooltip {...props}>
            {
                ({showTooltip, updateBounds}) => <>
                    {React.cloneElement(children as React.ReactElement, {
                        onLayout: ({nativeEvent}: LayoutChangeEvent) => {
                            updateBounds(nativeEvent.layout);
                            showTooltip();
                        },
                    })}
                </>
            }
        </CommonTooltip>
    );
}
```

Reply:
I have a few doubts, though:
- I'm not sure what hovering functionality is expected from the new "educational" tooltips. [I asked here](https://github.com/Expensify/App/issues/38054#issuecomment-2060812675).
  - Depending if we **want** hover support at all for them, different implementation strategies might be viable
- I'm not sure what is the precise characteristic of the layout handling you implemented for Native. You used cross-platform `onLayout` API for this. For the Web, we use a thrid-party `BoundsObserver` (which I'm the author of, by the way...) because as I remember it, we couldn't manage to achieve the desired functionality with `onLayout` only. Stuff broke when the "tooltiped" content was animating, when we were scrolling when the tooltip was visible, etc.
  - Anyway, I'm wondering if for the simpler case of the "educational" tooltip your `onLayout` approach can or can't work on Web too (especially desktop Web)
- The `CommonTooltip` I'm suggesting is named with a "common" prefix, not "base", more widespread in our codebase, only because we already have a class named `BaseTooltip`. It's implementation is quite trivial, though, and maybe it could be merged with the `CommonTooltip` I'm suggesting.

This is just a starting point, I covered `src/components/Tooltip/BaseTooltip/index.tsx` / `index.native.tsx` only for now. Other files can be de-duplicated using similar basic code abstraction techniques.

Reply:
>I'm not sure what hovering functionality is expected from the new "educational" tooltips. https://github.com/Expensify/App/issues/38054#issuecomment-2060812675.
> - Depending if we want hover support at all for them, different implementation strategies might be viable

Commented in the issue, hover support isn't a requirement. 👍

Reply:
Great! So I believe the strategy I sketched, modeling the `EducationalTooltip` as a simpler, cross-platform component without the "heavy" functionality of `HoverToopltip` might be the best way to go. We'll need to ensure things don't break during scrolling, though.

Reply:
Great! @tienifr does this make sense to you now, and can you prioritise working on the next steps today? Thanks!

Reply:
@tienifr 

Great progress so far!

`TooltipRenderedOnPageBody` is still basically a copy-pasted component, though. We should use the same techniques in this case.

As "page body" is a Web-specific concept, we could rename it to `RenderedTooltip` (I'm also open to suggestions).

Reply:
@trjExpensify Although there's great progress so far, I'd classify our task in PR as a complex one, because of the legacy of the existing tooltip component. We'll need more review rounds here if we want to put the whole scope in this PR.

Reply:
> How's about BaseGenericTooltip

Semantically, I don't think it's that's a lucky choice. In this case, we're not only having yet-another-layer of component that's actually used to render the same chunk of UI eventually. Remember that the name `Tootlip` was always meaning the tooltip target wrapper, while `TooltipRenderedOnPageBody` was focused on realizing the _actual_ tooltip wrapper & content.

![image](https://github.com/Expensify/App/assets/2590174/899ca4b8-8f99-4e92-9305-d7a56068e6fe)

Pragmatically, we have another problem. As I'm challenging the amount of duplication in this new `TooltipRenderedOnPageBody` replacement (let's call it `RenderedTooltip` for this reasoning), the _solution_ might involve creating `BaseRenderedTooltip`, while `BaseBaseGenericTooltip` sounds kind of mad.

So I'd vote for some _Base_-less _Generic_-less name, and `RenderedTooltip` sounded like a minimal change, just throwing out the DOM-specific aspect (`OnPageBody` /  `on page <body>`).

Reply:
What also somehow bothers me is that `BaseFoo` can mean anything, it just says "well, Foo was taken 🤷‍♂️". It's either because we want a `BaseFoo` optionally-rendered by `Foo` , it can also be cross-platform `BaseFoo`, where actual platform-specific `Foo` injects some delegates, and here it's just a `BaseFoo` rendered in another cooridante space, owned by `Foo`. It's a naming mess, in my opinion. I wouldn't add to it if possible.

Reply:
In our case, number 2 (_base foo_ = _cross-platform foo_) collides with number 3 (_base foo_ = _different coords foo_)

Reply:
@tienifr, we could at least lift this rendering logic up to a generic component, what do you think?
<img width="1135" alt="image" src="https://github.com/Expensify/App/assets/9059945/3977dd4b-5cf8-46ae-a8fb-b391047b7cc7">